crates/block-producer/src/mempool/dependency_graph.rs

```diff
@@ -0,0 +1,826 @@
+    MissingParents(BTreeSet<K>),
+
+    #[error("Nodes not found: {0:?}")]
+    UnknownNodes(BTreeSet<K>),
+
+    #[error("Nodes were not yet processed: {0:?}")]
+    UnprocessedNodes(BTreeSet<K>),
+
+    #[error("Nodes would be left dangling: {0:?}")]
+    DanglingNodes(BTreeSet<K>),
+
+    #[error("Node {0} is not ready to be processed")]
+    NotARootNode(K),
+}
+
+/// This cannot be derived without enforcing `Default` bounds on K and V.
+impl<K, V> Default for DependencyGraph<K, V> {
+    fn default() -> Self {
+        Self {
+            vertices: Default::default(),
+            parents: Default::default(),
+            children: Default::default(),
+            roots: Default::default(),
+            processed: Default::default(),
+        }
+    }
+}
+
+impl<K: Ord + Clone, V: Clone> DependencyGraph<K, V> {
+    /// Inserts a new node into the graph.
+    ///
+    /// # Errors
+    ///
+    /// Errors if the node already exists, or if any of the parents are not part of the graph.
+    ///
+    /// This method is atomic.
+    pub fn insert(&mut self, key: K, value: V, parents: BTreeSet<K>) -> Result<(), GraphError<K>> {
+        if self.vertices.contains_key(&key) {
+            return Err(GraphError::DuplicateKey(key));
+        }
+
+        let missing_parents = parents
+            .iter()
+            .filter(|parent| !self.vertices.contains_key(parent))
+            .cloned()
+            .collect::<BTreeSet<_>>();
+        if !missing_parents.is_empty() {
+            return Err(GraphError::MissingParents(missing_parents));
+        }
+
+        // Inform parents of their new child.
+        for parent in &parents {
+            self.children.entry(parent.clone()).or_default().insert(key.clone());
+        }
+        self.vertices.insert(key.clone(), value);
+        self.parents.insert(key.clone(), parents);
+        self.children.insert(key.clone(), Default::default());
+
+        self.try_make_root(key);
+
+        Ok(())
+    }
+
+    /// Reverts the nodes __and their descendents__, requeueing them for processing.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if any of the given nodes:
+    ///
+    ///   - are not part of the graph, or
+    ///   - were not previously processed
+    ///
+    /// This method is atomic.
+    pub fn revert_subgraphs(&mut self, keys: BTreeSet<K>) -> Result<(), GraphError<K>> {
+        let missing_nodes = keys
+            .iter()
+            .filter(|key| !self.vertices.contains_key(key))
+            .cloned()
+            .collect::<BTreeSet<_>>();
+        if !missing_nodes.is_empty() {
+            return Err(GraphError::UnknownNodes(missing_nodes));
+        }
+        let unprocessed = keys.difference(&self.processed).cloned().collect::<BTreeSet<_>>();
+        if !unprocessed.is_empty() {
+            return Err(GraphError::UnprocessedNodes(unprocessed));
+        }
+
+        let mut processed = BTreeSet::new();
+        let mut to_process = keys.clone();
```

Comment:
nit: `processed` and `to_process` naming here is a bit confusing because we also have `self.processed`. Maybe `reverted` and `to_revert` would be better naming?